src/aktualizr_secondary/aktualizr_secondary.cc

```diff
@@ -128,23 +130,46 @@ bool AktualizrSecondary::sendFirmwareResp(const std::shared_ptr<std::string>& fi
 #endif
   } else if (pacman->name() == "debian") {
     // TODO save debian package here.
-    LOG_ERROR << "Installation of non ostree images is not suppotrted yet.";
+    LOG_ERROR << "Installation of debian images is not suppotrted yet.";
     return false;
   }
 
-  // This check is not applicable for the ostree case, just for the binary file case(s)
-  if (targetToApply.length() != firmware->length()) {
+  if (targetToApply.length() != firmware_size) {
     LOG_ERROR << "The target image size specified in metadata " << targetToApply.length()
               << " does not match actual size " << firmware->length();
     return false;
   }
 
+  if (!targetToApply.IsOstree()) {
```

Comment:
That's odd - why do we have to brute-force hash type in here?

Reply:
I am not sure if I understand the question.

Reply:
Why both sha256 and sha512 are computed simultaneously? We don't know which hash is actually used?

Reply:
As a matter of fact, the targets.json sent from BE contains both type of hashes for each "BINARY" target format.
Also, primary supports these two types of hashes and can generate and verify any of them, although it verifies just the first hash found in json :). See the following for more details 

https://github.com/advancedtelematic/aktualizr/blob/6b1da3e473f3c9963a3221607dabc4d0cde06968/src/libaktualizr/package_manager/packagemanagerinterface.cc#L149

https://github.com/advancedtelematic/aktualizr/blob/6b1da3e473f3c9963a3221607dabc4d0cde06968/src/libaktualizr/package_manager/packagemanagerinterface.cc#L8

The questions are 
1) why there are two types of hashes present in targets.json ? 
2) why Primary verifies just the first found hash for a binary image instead of verifying each hash ?
3) what Secondary should do in order to support Full Uptane Verification ?

Reply:
That's weird indeed - I can't really picture situation when sha256 will match but sha512 wouldn't or other way around. @patrickvacek, @lbonn any ideas what's going on?

Reply:
Well, if one matches and the other doesn't, either somebody is doing a weird attack (and if they managed to sign the metadata, they can try more obvious things) or one of the hash function has been broken.

For 1., that's something that's allowed by Uptane, supposedly for hash algorithms agility, so I suppose that's "because we can". But then, should we really?..

For 2., I suppose it's technically a bug then.

Reply:
Makes sense - it's clearly a bug if we only check one hash instead of all available (and we should fail if any of them mismatch) as it defeats the purpose of having multiple hashes in a first place. Do we already have a ticket for this?

Reply:
> Do we already have a ticket for this?
No, I can create one so we can discuss a need in multiple hashes and their verification in scope of that ticket.

@zabbal  @lbonn What do you think I should do specifically in this pull request ?

Leave it as it is now
Mimic the current primary's logic, i.e. get the first found hash in the target,json, generate corresponding hash for received data and check whether the hashes matches

I/We can change the hash verification logic for both Primary and Secondary once we discuss and agree on it in the scope of another ticket and a pull request, just would like to unblock merging of this PR.

Reply:
I've just double check this 'stuff' and it turned out that I was wrong when I was saying that there are two hashes in target.json for a binary image file. It turned out that the BE in the prod actually sends just one hash - sha256
 
`"targets": {
      "whatever_0.1": {
        "hashes": {
          "sha256": "30e14955ebf1352266dc2ff8067e68104607e750abb9d3b36582b8af909fcb58"
        },
`
The reason I was the original statement about the two hashes is the fact that 'uptane-genertor' actually generates two type of hashes

`"update_100.0": {
        "custom": {
          "ecuIdentifiers": {
            "some-serial": {
              "hardwareId": "secondary-ecu-hdw-ID-001"
            }
          },
          "targetFormat": "BINARY"
        },
        "hashes": {
          "sha256": "a866dd69dc84edc2991085f7a36bbd59f473ce4c8c18557369891bb5d4807934",
          "sha512": "d4f562d1d23282b70a0d67a25474b632f184321f5fd0abede9f4b0ae1ef2c260ef43ea4b06e83b28d05dcb527423156049247f99e978bd4633ffb6ab228fd736"
        },`


Therefore, I am going to apply to Secondary the same hash verification logic as in the case of Primary, which is check the first found hash in targets.json.